src/audio/chain_dma.c

```diff
@@ -37,56 +48,478 @@ static struct comp_dev *chain_dma_create(const struct comp_driver *drv,
+		return IPC4_FAILURE;
+	}
+
+	fifo_size = ALIGN_UP_INTERNAL(fifo_size, addr_align);
+
+	if (cd->link_connector_node_id.f.dma_type == ipc4_hda_link_input_class) {
+		/* Increasing buffer size for capture path as L1SEN exit takes sometimes
+		 * more than expected.
+		 */
+		fifo_size *= 5;
+	}
+
+	if (cd->link_connector_node_id.f.dma_type == ipc4_hda_link_output_class) {
+		/* Increasing buffer size for playback path as L1SEN exit takes sometimes
+		 * more that expected
+		 * Note, FIFO size must be smaller than half of host buffer size
+		 * (20ms ping pong) to avoid problems with position reporting
+		 * Size increase from default 2ms to 5ms is enough.
+		 */
+		fifo_size /= 2;
+		fifo_size *= 5;
+	}
+
+	cd->dma_buffer = buffer_alloc(fifo_size, SOF_MEM_CAPS_DMA,
+				      addr_align);
+
+	if (!cd->dma_buffer) {
+		comp_err(dev, "chain_task_init(): failed to alloc dma buffer");
+		return IPC4_FAILURE;
+	}
+
+	/* clear dma buffer */
+	buffer_c = buffer_acquire(cd->dma_buffer);
+	buffer_zero(buffer_c);
+	buffer_release(buffer_c);
+
+	ret = chain_init(dev, cd->dma_buffer->stream.addr,
+			 cd->dma_buffer->stream.size);
+	if (ret)
+		return IPC4_FAILURE;
+
+	cd->chain_task.state = SOF_TASK_STATE_INIT;
+
+	return IPC4_SUCCESS;
+}
+
+int chain_init(struct comp_dev *dev, void *addr, size_t length)
+{
+	struct chain_dma_data *cd = comp_get_drvdata(dev);
+	struct dma_config *dma_cfg_host = &cd->z_config_host;
+	struct dma_config *dma_cfg_link = &cd->z_config_link;
+	struct dma_block_config *dma_block_cfg_host = &cd->dma_block_cfg_host;
+	struct dma_block_config *dma_block_cfg_link = &cd->dma_block_cfg_link;
+
+	memset(dma_cfg_host, 0, sizeof(*dma_cfg_host));
+	memset(dma_cfg_link, 0, sizeof(*dma_cfg_link));
+	memset(dma_block_cfg_host, 0, sizeof(*dma_block_cfg_host));
+	memset(dma_block_cfg_link, 0, sizeof(*dma_block_cfg_link));
+
+	dma_cfg_host->block_count = 1;
+	dma_cfg_host->source_data_size = cd->cs;
+	dma_cfg_host->dest_data_size = cd->cs;
+	dma_cfg_host->head_block = dma_block_cfg_host;
+	dma_cfg_link->block_count = 1;
+	dma_cfg_link->source_data_size = cd->cs;
+	dma_cfg_link->dest_data_size = cd->cs;
+	dma_cfg_link->head_block  = dma_block_cfg_link;
+
+	dma_block_cfg_host->block_size = length;
+	dma_block_cfg_link->block_size = length;
+
+	switch (cd->stream_direction) {
+	case SOF_IPC_STREAM_PLAYBACK:
+		dma_cfg_host->channel_direction = HOST_TO_MEMORY;
+		dma_block_cfg_host->dest_address = (uint32_t)addr;
+		dma_cfg_link->channel_direction = MEMORY_TO_PERIPHERAL;
+		dma_block_cfg_link->source_address = (uint32_t)addr;
+		break;
+	case SOF_IPC_STREAM_CAPTURE:
+		dma_cfg_host->channel_direction = MEMORY_TO_HOST;
+		dma_block_cfg_host->source_address = (uint32_t)addr;
+		dma_cfg_link->channel_direction = PERIPHERAL_TO_MEMORY;
+		dma_block_cfg_link->dest_address = (uint32_t)addr;
+		break;
+	}
+
+	/* get host DMA channel */
+	int channel = cd->host_connector_node_id.f.v_index;
+
+	channel = dma_request_channel(cd->dma_host->z_dev, &channel);
```

Comment:
It would be easier to follow if you do:
```c
	channel = dma_request_channel(cd->dma_host->z_dev,
				      &cd->host_connector_node_id.f.v_index);
...
	channel = dma_request_channel(cd->dma_host->z_dev,
				      &cd->link_connector_node_id.f.v_index);
```

Reply:
I see... but compiler cannot take address of bit-field 'v_index', can we stay with anything like previous approach with declaration on the beginning ?